docs/posts/pydata-performance/index.qmd

```diff
@@ -0,0 +1,696 @@
+
+This worked and the workload completed. However, after talking to Dask
+expert and Ibis contributor [Naty Clementi](https://github.com/ncclementi) she
+suggested I try a few things:
+
+* Use [the distributed scheduler](https://distributed.dask.org/en/stable/).
+* Ensure that [`pyarrow` string arrays are
+  used](https://docs.dask.org/en/latest/configuration.html#dask) instead of
+  NumPy object arrays.
+
+Both of those settings together shaved about 2x off the
+non-distributed-scheduler, non-pyarrow-string version while adding only
+a couple lines of code. Nice!
+
+Let's look at the Dask implementation:
+
+```{python}
+#| echo: false
+#| output: asis
+with open("./dask_impl.py") as f:
+    print(f"```python\n{f.read()}\n```")
+```
+
+Let's run the code:
+
+```{python}
+dask_results = %timeit -n1 -r1 -o %run dask_impl.py
+```
+
+That's a nice improvement over Pandas: we finished the workload. Dask is still
+good ways behind DuckDB out of the box!
+
+## Takeaways
+
+**Ibis + DuckDB is the only system tested that handles this workload well out of the box**
+
+Pandas couldn't handle the workload due to memory constraints, and Dask
+required some fairly advanced knowledge to get maximum performance.
+
+Let's recap the results with some numbers:
+
+### Numbers
+
+```{python}
+#| echo: false
+#| output: asis
+smallest_file_size = os.path.getsize(smallest_file)
+total_size = sum(map(os.path.getsize, allfiles))
+
+
+def make_line(name, results):
+    duration = results.best
+    total_size_mib = total_size / (1 << 20)
+    throughput = round(total_size_mib / duration, 0)
+    return " | ".join(
+        [
+            name,
+            f"{int(round(total_size_mib, 0)):,} MiB",
+            f"{int(round(duration, 0))} seconds",
+            f"{int(throughput)} MiB/s"
+        ]
+    )
+
+
+header = "| Toolset | Data size | Duration | Throughput |"
+sep = "| ------------------ | --------: | -----------: | ---------: |"
+rows = [
+    header,
+    sep,
+    f"| {make_line('Ibis + DuckDB', duckdb_results)} |",
+    f"| {make_line('Dask + Distributed', dask_results)} |",
+]
+print("\n".join(rows))
+```
+
+With Ibis + DuckDB, I was able to write the query the way I wanted to without
+running out of memory, using the default configuration provided by Ibis.
+
+```{python}
+#| echo: false
+#| output: asis
+pandas_duration = pandas_results.best
+pandas_expected_duration = pandas_duration * (total_size / smallest_file_size)
+duckdb_duration = duckdb_results.best
+speedup = int(round(pandas_expected_duration / duckdb_duration, 0))
+
+print(
+    f"I was able execute this computation about **{speedup}x faster** than you can "
+    "expect with pandas."
+)
+```
+
+In contrast, Pandas ran out of memory **on a single file** without some hand
+holding, and while Dask didn't cause the Linux OOM-killer to start terrorizing
+my program it took about 4x longer than DuckDB, used a lot more memory and
+required configuration to get the best performance.
```

Comment:
What was the configuration you're referring to here?  The use of `Client()`?  If so, this feels analagous to DuckDB's creation of a connection object (although maybe that's unnecessary with Ibis)

Reply:
Fair point about `Client` vs connection.

We support calling `ibis.read_parquet` without the need for an explicit connection.

Here are main my complaints:

1. The distributed scheduler isn't the default :)
2. I have to be aware of `fork` behavior when using it (sticking my code under a main check)
3. I have to explicitly tell it to shutdown workers (to get them to release memory). I understand a context manager is an option, but now I have another level of nesting.

Reply:
Yeah, I think I agree with the confusion around processes-in-processes and `if __name__ == "__main__":`, that's certainly a valid concern.

I guess when I hear people talk about "required configuration" I assume they mean "I had to research and set a bunch of options in a config file" similar to the historical complaints with Spark or similar systems.  This feels different from that.  Maybe "used a lot more memory and required me to create a `Client()` object to get the best performance" gives a more faithful description?

Not a big deal though.  This is just splitting hairs.

Reply:
Also, maybe worth noting, but all of the dask examples tend to start with

```python
from dask.distributed import Client
client = Client()
```